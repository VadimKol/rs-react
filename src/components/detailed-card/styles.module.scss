.detailed_Card {
  display: flex;
  flex-direction: column;
  gap: 20px;
  align-items: center;
}

.card {
  position: relative;
  width: 265px;
  background:
    linear-gradient(var(--background-color) 0 0) padding-box,
    linear-gradient(to top, var(--primary-color), var(--secondary-color)) border-box;
  border: 5px solid transparent;
  border-radius: 10px;
}

.text_container {
  display: flex;
  flex-direction: column;
  gap: 10px;
  padding: 20px 30px;
  box-shadow: inset 0 5px 5px 0 #d3d3d3;
}

.card_title {
  margin: 0;
  font-size: 18px;
  font-weight: 800;
  text-align: center;
}

.desc {
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.desc_item,
.type {
  margin: auto;
}

.episodes {
  display: -webkit-box;
  overflow: hidden;
  margin: 0 auto;
  text-align: left;
  text-overflow: ellipsis;
  -webkit-line-clamp: 5;
  -webkit-box-orient: vertical;
}

.close {
  font-size: 1.3rem;
}

.loader {
  width: 265px;
  height: 265px;
  background-image: url("/src/assets/images/loader.png");
  background-position: center;
  background-size: contain;
  background-repeat: no-repeat;
  animation: loader 1s infinite linear;
}

@keyframes loader {
  to {
    transform: rotate(1turn);
  }
}
