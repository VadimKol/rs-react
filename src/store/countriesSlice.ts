import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { countries } from '@/common/countries';

import { type RootState } from './store';

export const countriesLice = createSlice({
  name: 'countries',
  initialState: countries,
  reducers: {
    addCountry: (state, action: PayloadAction<string>) => {
      state.push(action.payload);
    },
  },
});

export const { addCountry } = countriesLice.actions;

export const selectCountries = (state: RootState): string[] => state.countries;

export const useCountries = (): string[] => useSelector(selectCountries);
