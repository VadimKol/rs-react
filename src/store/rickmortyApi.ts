import { createApi, fakeBaseQuery } from '@reduxjs/toolkit/query/react';
import { getCharacter, getCharacters } from 'rickmortyapi';

export const rickmortyApi = createApi({
  reducerPath: 'rickmortyApi',
  baseQuery: fakeBaseQuery(),
  endpoints: (builder) => ({
    getCharacters: builder.query({
      queryFn: async ({ page, character }: { page: number; character: { name: string } }) => {
        try {
          const { data, status, statusMessage } = await getCharacters({ page, name: character.name });

          if (status === 200) {
            return { data: { characters: data.results || [], totalPages: data.info?.pages || 0 } };
          }
          if (status === 404) {
            return { data: { characters: [], totalPages: 0 } };
          }

          throw new Error(statusMessage);
        } catch (error) {
          return { error };
        }
      },
    }),
    getCharacter: builder.query({
      queryFn: async (id: number) => {
        try {
          const { data, status, statusMessage } = await getCharacter(id);

          if (status === 200) {
            return { data };
          }
          if (status === 404) {
            return { data: null };
          }

          throw new Error(statusMessage);
        } catch (error) {
          return { error };
        }
      },
    }),
  }),
});

export const { useGetCharactersQuery, useGetCharacterQuery } = rickmortyApi;
