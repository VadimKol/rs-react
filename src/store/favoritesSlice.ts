/* import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import { type Character } from 'rickmortyapi';

import { type RootState } from './store';

const initialState: Character[] = [];

export const favoritesSLice = createSlice({
  name: 'favorites',
  initialState,
  reducers: {
    addToFavorites: (state, action: PayloadAction<Character>) => {
      state.push(action.payload);
    },
    removeFromFavorites: (state, action: PayloadAction<Character>) =>
      state.filter((character) => character.id !== action.payload.id),
    clearFavorites: () => [],
  },
});

export const { addToFavorites, removeFromFavorites, clearFavorites } = favoritesSLice.actions;

const selectFavorites = (state: RootState): Character[] => state.favorites;

const isFavorite = (state: RootState, character?: Character): boolean =>
  state.favorites.some((favorite) => favorite.id === character?.id);

export const useFavorites = (): Character[] => useSelector(selectFavorites);

export const useIsFavorite = (character?: Character): boolean =>
  useSelector((state: RootState) => isFavorite(state, character));
 */
