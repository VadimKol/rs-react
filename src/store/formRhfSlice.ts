import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

const initialState: FormInfo = {
  name: null,
  age: null,
  email: null,
  password: null,
  confirmPassword: null,
  gender: null,
  country: null,
  tc: null,
  image: null,
  strength: '0',
};

export const formRhfSLice = createSlice({
  name: 'formRhf',
  initialState,
  reducers: {
    updateFormRhf: (_, action: PayloadAction<FormInfo>) => action.payload,
  },
});

export const { updateFormRhf } = formRhfSLice.actions;

const selectFormRhfInfo = (state: RootState): FormInfo => state.formRhf;

export const useFormRhfInfo = (): FormInfo => useSelector(selectFormRhfInfo);
