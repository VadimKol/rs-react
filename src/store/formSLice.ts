import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

export const initialState: FormInfo[] = [];

export const formSLice = createSlice({
  name: 'form',
  initialState,
  reducers: {
    updateForm: (state, action: PayloadAction<FormInfo>) => {
      state.push(action.payload);
    },
  },
});

export const { updateForm } = formSLice.actions;

const selectFormInfo = (state: RootState): FormInfo[] => state.form;

export const useFormInfo = (): FormInfo[] => useSelector(selectFormInfo);
