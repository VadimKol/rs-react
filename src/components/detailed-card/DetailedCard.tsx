import { useRouter } from 'next/router';
import { type ReactNode } from 'react';

import { CustomButton } from '@/components/custom-button/Ð¡ustomButton';
import { FavoriteButton } from '@/components/favorite-button/FavoriteButton';
import { ImageBlock } from '@/components/image-block/ImageBlock';
import { useGetCharacterQuery } from '@/store/rickmortyApi';

import styles from './styles.module.scss';

export default function DetailedCard({ characterID }: { characterID: string }): ReactNode {
  const { query, push, replace } = useRouter();
  const { data: character, isFetching: loader, isError, error } = useGetCharacterQuery(Number(characterID));

  if (isError) {
    throw error;
  }

  if (character === null) {
    replace('*').catch(() => {});
    return null;
  }

  const desc = [
    `Species: ${character?.species}`,
    `Gender: ${character?.gender}`,
    `Status: ${character?.status}`,
    `Origin: ${character?.origin?.name}`,
    `Location: ${character?.location?.name}`,
  ];

  return (
    <section className={styles.detailed_Card}>
      {loader ? (
        <div className={styles.loader} />
      ) : (
        <>
          <div className={styles.card}>
            <ImageBlock src={character?.image || ''} alt="Character" />
            <div className={styles.text_container}>
              <h2 className={styles.card_title}>{character?.name}</h2>
              <ul className={styles.desc}>
                {desc.map((item) => (
                  <li key={item} className={styles.desc_item}>
                    {item}
                  </li>
                ))}
              </ul>
              {character?.type !== '' && <p className={styles.type}>Type: {character?.type}</p>}
              <p className={styles.episodes}>
                Episodes: {character?.episode.map((episode) => episode.replace(/\D/g, '')).join(', ')}
              </p>
              <FavoriteButton character={character} />
            </div>
          </div>
          <CustomButton
            className={styles.close}
            onClick={() => {
              push({ pathname: '/', query: { page: query.page } }).catch(() => {});
            }}
          >
            Close
          </CustomButton>
        </>
      )}
    </section>
  );
}
