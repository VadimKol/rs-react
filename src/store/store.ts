import { configureStore } from '@reduxjs/toolkit';

import { formSLice } from './formSLice';

export const store = configureStore({ reducer: { form: formSLice.reducer } });

export type RootState = ReturnType<typeof store.getState>;
