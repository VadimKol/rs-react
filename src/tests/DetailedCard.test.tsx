import { createRemixStub } from '@remix-run/testing';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { type ReactNode } from 'react';
import { Provider } from 'react-redux';

import { DetailedCard } from '@/components/detailed-card/DetailedCard';
import { store } from '@/store/store';

import { character } from './mocks/data';

const navigate = vi.fn((arg: string): string => arg);

describe('DetailedCard Component', () => {
  const user = userEvent.setup();

  it('renders correctly', () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: (): ReactNode => (
          <Provider store={store}>
            <DetailedCard character={character} />
          </Provider>
        ),
      },
    ]);

    const { container } = render(<RemixStub />);

    expect(container).toMatchSnapshot();
  });

  it('changes URL when click on Close button', async () => {
    vi.mock('@remix-run/react', () => ({
      useSearchParams: vi.fn(() => [new URLSearchParams('')]),
      useNavigate: vi.fn(() => navigate),
    }));

    render(
      <Provider store={store}>
        <DetailedCard character={character} />
      </Provider>,
    );

    const buttonSearch = screen.getByText('Close');

    await user.click(buttonSearch);

    expect(navigate).toHaveBeenCalledWith(`/?page=1&search=`);
  });

  it('did not render when person is not specified', () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: (): ReactNode => (
          <Provider store={store}>
            <DetailedCard character={null} />
          </Provider>
        ),
      },
    ]);

    const { container } = render(<RemixStub />);

    expect(container).toBeEmptyDOMElement();
  });
});
