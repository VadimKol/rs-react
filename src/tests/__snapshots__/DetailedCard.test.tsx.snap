// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DetailedCard Component > renders correctly 1`] = `
<div>
  <section
    class="_detailed_Card_efa844"
  >
    <div
      class="_card_efa844"
    >
      <div
        class="_image_container_de2c00"
      >
        <div
          class="_loader_d8e6ec"
        />
        <img
          alt="Character"
          class="_hidden_de2c00"
          height="300"
          src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
          width="300"
        />
      </div>
      <div
        class="_text_container_efa844"
      >
        <h2
          class="_card_title_efa844"
        >
          Rick
        </h2>
        <ul
          class="_desc_efa844"
        >
          <li
            class="_desc_item_efa844"
          >
            Species: 
          </li>
          <li
            class="_desc_item_efa844"
          >
            Gender: Male
          </li>
          <li
            class="_desc_item_efa844"
          >
            Status: Alive
          </li>
          <li
            class="_desc_item_efa844"
          >
            Origin: 
          </li>
          <li
            class="_desc_item_efa844"
          >
            Location: 
          </li>
        </ul>
        <p
          class="_type_efa844"
        >
          Type: 
          Human
        </p>
        <p
          class="_episodes_efa844"
        >
          Episodes: 
        </p>
        <button
          aria-label="Add to favorites"
          class="_favorite_99b3bf"
          name="favorite"
          type="button"
          value="false"
        >
          â˜†
        </button>
      </div>
    </div>
    <button
      aria-label="Button"
      class="_button_de8d5c _close_efa844"
      id=""
      type="button"
    >
      Close
    </button>
  </section>
</div>
`;
